README.md
# taalika: plain text tables in Rust

[![Crates.io](https://img.shields.io/crates/v/taalika.svg?maxAge=2592000)](https://crates.io/crates/taalika)
[![Documentation (latest release)](https://docs.rs/taalika/badge.svg)](https://docs.rs/taalika/)
[![Documentation (main)](https://img.shields.io/badge/docs-main-59f)](https://sunshowers-code.github.io/taalika/rustdoc/taalika/)
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE-MIT)
[![License: Apache 2.0](https://img.shields.io/badge/license-Apache_2.0-blue.svg)](LICENSE-APACHE)

Builds plain, automatically-aligned tables of monospaced text.
This is basically what you want if you are implementing `ls`.

## Example

```rust
use taalika::{Table, Row};
use std::path::Path;

fn ls(dir: &Path) -> ::std::io::Result<()> {
    let mut table = Table::new("{:>}  {:<}{:<}  {:<}");
    for entry_result in ::std::fs::read_dir(dir)? {
        let entry    = entry_result?;
        let metadata = entry.metadata()?;

        table.add_row(Row::new()
             .with_cell(metadata.len())
             .with_cell(if metadata.permissions().readonly() {"r"} else {""})
             .with_cell(if metadata.is_dir() {"d"} else {""})
             .with_cell(entry.path().display()));
    }

    print!("{}", table);

    Ok(())
}

ls(Path::new("target")).unwrap();
```

produces something like

```
1198     target/.rustc_info.json
1120  d  target/doc
 192  d  target/package
1056  d  target/debug
```

## Other features

  - The `Table::with_heading()` and `Table::add_heading()` methods add
    lines that span all columns.

  - The `row!` macro builds a row with a fixed number of columns
    using less syntax.

  - The `Table::set_line_end()` method allows changing the line ending
    to include a carriage return (or whatever you want).

  - The `Row::with_ansi_cell` and `Row::add_ansi_cell` methods can be
    used to add cells with ANSI color codes, and still have their widths be
    computed correctly.


## Usage

`taalika` is on [crates.io](https://crates.io/crates/taalika).

## Features

* Feature `unicode-width` is used to compute the width of columns in terms of
  Unicode graphemes. It is enabled by default and depends on the
  [unicode-width](https://crates.io/crates/unicode-width) crate.

  Note that without `unicode-width`, alignment will be based on the count of the
  `std::str::Chars` iterator.

This crate supports Rust version 1.46.0 and later.

## See also

You may also want:

- [tabular](https://crates.io/crates/tabular) - `taalika` is a fork of this
  crate with additional features.

- [text-tables](https://crates.io/crates/text-tables) – This is more automatic
  than tabular. You give it an array of arrays, it renders a nice table with 
  borders. Tabular doesn't do borders.

- [prettytable](https://crates.io/crates/prettytable-rs) — This has an API more
  similar to tabular’s in terms of building a table, but it does a lot more, 
  including, color, borders, and CSV import.


CHANGELOG.md
# Changelog

This project adheres to [Semantic Versioning].

[Semantic Versioning]: http://semver.org/spec/v2.0.0.html

## [0.1.6] - 2021-12-21

### Changed

- Minor documentation updates.

## [0.1.5] - 2021-12-21

### Added

- `Row::with_ansi_cell` and `add_ansi_cell` add cells that potentially have
  ANSI color codes in them; the color codes are stripped out while computing
  the width.
- `Row::with_custom_width_cell` and `add_custom_width_cell` add cells with a
  custom width.

### Changed

- Project forked and renamed to `taalika`.

---

The following releases were for the tabular crate, which this is a fork of.

## [0.1.4] - 2019-12-29

### Changed
- Oldest supported Rust version is now 1.31.0.

## [0.1.3] - 2019-12-29

### Changed
- No longer depends on deprecated `str::trim_right` method.

## [0.1.2] - 2018-09-18

### Added
- `Table::set_line_end()` method for changing the line ending used by
formatted tables.
- `row!()` and `table!()` macros.

### Changed
- Centering now rounds to left rather than right; I think it looks better.

## [0.1.1] - 2018-09-18

### Added
- `{:^}` column specifier for centering.

## [0.1.0] - 2019/09/18

Initial release.

[0.1.6]: https://github.com/sunshowers-code/taalika/releases/tag/0.1.6
[0.1.5]: https://github.com/sunshowers-code/taalika/releases/tag/0.1.5


src/lib.rs
// Copyright (c) taalika Contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

//! Builds plain, automatically-aligned tables of monospaced text.
//! This is basically what you want if you are implementing `ls`.
//!
//! # Example
//!
//! ```
//! use taalika::{Table, Row};
//! use std::path::Path;
//!
//! fn ls(dir: &Path) -> ::std::io::Result<()> {
//!     let mut table = Table::new("{:>}  {:<}{:<}  {:<}");
//!     for entry_result in ::std::fs::read_dir(dir)? {
//!         let entry    = entry_result?;
//!         let metadata = entry.metadata()?;
//!
//!         table.add_row(Row::new()
//!              .with_cell(metadata.len())
//!              .with_cell(if metadata.permissions().readonly() {"r"} else {""})
//!              .with_cell(if metadata.is_dir() {"d"} else {""})
//!              .with_cell(entry.path().display()));
//!     }
//!
//!     print!("{}", table);
//!
//!     Ok(())
//! }
//!
//! ls(Path::new("target")).unwrap();
//! ```
//!
//! produces something like
//!
//! ```text
//! 1198     target/.rustc_info.json
//! 1120  d  target/doc
//!  192  d  target/package
//! 1056  d  target/debug
//! ```
//!
//! # Other features
//!
//!   - The [`Table::with_heading`] and [`Table::add_heading`] methods add
//!     lines that span all columns.
//!
//!   - The [`row!`] macro builds a row with a fixed number of columns
//!     using less syntax.
//!
//!   - The [`Table::set_line_end`] method allows changing the line ending
//!     to include a carriage return (or whatever you want).
//!
//!   - The [`Row::with_ansi_cell`] and [`Row::add_ansi_cell`] methods can be
//!     used to add cells with ANSI color codes, and still have their widths be
//!     computed correctly.
//!
//! # Usage
//!
//! `taalika` is on [crates.io](https://crates.io/crates/taalika).
//!
//! # Features
//!
//! * Feature `unicode-width` is used to compute the width of columns in terms of
//!   Unicode graphemes. It is enabled by default and depends on the
//!   [unicode-width](https://crates.io/crates/unicode-width) crate.
//!
//!   Note that without `unicode-width`, alignment will be based on the count of the
//!   `std::str::Chars` iterator.
//!
//! This crate supports Rust version 1.46.0 and later.
//!
//! # See also
//!
//! You may also want:
//!
//! - [tabular](https://crates.io/crates/tabular) - `taalika` is a fork of this
//!   crate with additional features.
//!
//! - [text-tables](https://crates.io/crates/text-tables) – This is more automatic
//!   than tabular. You give it an array of arrays, it renders a nice table with
//!   borders. Tabular doesn't do borders.
//!
//! - [prettytable](https://crates.io/crates/prettytable-rs) — This has an API more
//!   similar to tabular’s in terms of building a table, but it does a lot more,
//!   including, color, borders, and CSV import.
//!
//! [`row!`]: macro.row.html

#![warn(missing_docs)]

mod column_spec;
mod error;
mod macros;
mod row;
mod table;
mod width_string;

pub use crate::{
    error::{Error, Result},
    row::Row,
    table::Table,
};

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn alignment() {
        let mut table = Table::new("{:>}  ({:<}) {:<}");
        table
            .add_row(Row::new().with_cell(1).with_cell("I").with_cell("one"))
            .add_row(Row::new().with_cell(5).with_cell("V").with_cell("five"))
            .add_row(Row::new().with_cell(10).with_cell("X").with_cell("ten"))
            .add_row(Row::new().with_cell(50).with_cell("L").with_cell("fifty"))
            .add_row(
                Row::new()
                    .with_cell(100)
                    .with_cell("C")
                    .with_cell("one-hundred"),
            );
        assert_eq!(
            format!("\n{}", table),
            r#"
  1  (I) one
  5  (V) five
 10  (X) ten
 50  (L) fifty
100  (C) one-hundred
"#
        );
    }

    #[test]
    fn heading() {
        let _row = Row::from_cells(vec!["a", "b", "c"]);
        //        eprintln!("{:?}", _row);

        let table = Table::new("{:<} {:<} {:>}")
            .with_row(Row::from_cells(vec!["a", "b", "d"]))
            .with_heading("This is my table")
            .with_row(Row::from_cells(vec!["ab", "bc", "cd"]));

        //        eprintln!("\n\n{:?}\n\n", table);

        assert_eq!(
            format!("\n{}", table),
            r#"
a  b   d
This is my table
ab bc cd
"#
        );
    }

    #[test]
    fn centering() {
        let table = Table::new("{:<} {:^} {:>}")
            .with_row(Row::from_cells(vec!["a", "b", "c"]))
            .with_row(Row::from_cells(vec!["a", "bc", "d"]))
            .with_row(Row::from_cells(vec!["a", "bcd", "e"]))
            .with_row(Row::from_cells(vec!["a", "bcde", "f"]))
            .with_row(Row::from_cells(vec!["a", "bcdef", "g"]));

        assert_eq!(
            format!("\n{}", table),
            r#"
a   b   c
a  bc   d
a  bcd  e
a bcde  f
a bcdef g
"#
        );
    }

    #[test]
    fn temporary() {}
}
